### PR TITLE
Fixed a bug where ipy notebooks were split across test groups

### LINKED ISSUE #9: Group IPython notebooks as a single file
Currently, each cell is treated as an individual unit, but that quickly breaks as objects created in previous cells are not accessible in the rest of the notebook (plus imports).

### PR DESCRIPTION
**Pull request by ma-sadeghi**
Fixes #9. @jerry-git Could you please take a look and merge if you agree? Thanks!

### CODE REVIEW

File: src/pytest_split/plugin.py [hunk 1]
```diff
@@ -134,4 +134,37 @@ def _calculate_suite_start_and_end_idx(splits, group, items, stored_durations):
...
+        end_idx = _get_boundary_idx_for_ipynb(item_node_ids, end_idx, "end")
+
     return start_idx, end_idx
+
+
+def _is_ipy_notebook(node_id):
+    path = node_id.split("::")[0]
+    if path.endswith(".ipynb"):
+        return True
+    return False
```
Comment (ma-sadeghi) [inline src/pytest_split/plugin.py:150]: This helper function is for inferring whether a node_id is related to an IPython notebook or not.

File: src/pytest_split/plugin.py [hunk 2]
```diff
@@ -134,4 +134,37 @@ def _calculate_suite_start_and_end_idx(splits, group, items, stored_durations):
     if not end_idx:
         end_idx = len(items)
 
+    # Modify start_idx and end_idx so they match start/end of an ipynb file
+    if _is_ipy_notebook(item_node_ids[start_idx]):
+        start_idx = _get_boundary_idx_for_ipynb(item_node_ids, start_idx, "start")
+    if _is_ipy_notebook(item_node_ids[end_idx - 1]):  # since Python indexing is [a, b)
+        end_idx = _get_boundary_idx_for_ipynb(item_node_ids, end_idx, "end")
```
Comment (ma-sadeghi) [inline src/pytest_split/plugin.py:141]: Here, I will first inspect whether the node_id associted with `start_idx` and `end_idx` are IPython notebooks or not, and if so, I modify the indices so a single IPython notebook is not broken into different test groups.

File: src/pytest_split/plugin.py [hunk 3]
```diff
@@ -134,4 +134,37 @@ def _calculate_suite_start_and_end_idx(splits, group, items, stored_durations):
...
+    while same_notebook:
+        if idx in [0, len(item_node_ids)]:
+            break
+        idx_next = idx - 1 if mode == "start" else idx + 1
+        fname_next = item_node_ids[idx_next].split("::")[0]
+        if fname_next == fname:
+            idx = idx_next
+        else:
+            same_notebook = False
+    return idx if mode == "start" else idx + 1  # since Python indexing is [a, b)
```
Comment (ma-sadeghi) [inline src/pytest_split/plugin.py:170]: This is a helper function for obtaining the start/end indices of an IPython notebook, given node_id corresponding to a single cell of that file.

**Comment by jerry-git** [comment]
Hello @ma-sadeghi! Great to hear that you're using `pytest-split`!

Could you briefly explain what's your use case with iPython notebooks, tests, and pytest-split? If you can share some example notebook(s), even better.

I'm just a bit skeptical whether a pytest plugin itself is the correct place for notebook support 🤔  I mean, it's quite a bit of added complexity to have around 25% of all the code just for IPython notebook support. Perhaps you could fork this and create e.g. `pytest-split-ipynb`?

**Comment by ma-sadeghi** [comment]
Thanks for getting back to me.

Let me back up. So, there's this plugin [`nbval`](https://github.com/computationalmodelling/nbval) which allows testing IPython notebooks along with good old pytest unit tests via
```
pytest --nbval test_folder
```
In our packages, we use a combination of unit tests and IPython notebooks for testing (unit tests just for testing and notebooks both for testing and tutorial for users). Since we have many of these, we want to parallelize the testing process. But, the issue is that `pytest-split` breaks the notebooks, because for instance, let's say you define `x`
```
x = 1
``` 
in the first cell of the notebook, and use the value of `x` in another cell
```
y = x + 1
```
in this simple example, in principle `pytest-split` could put cell 1 in group 1 and cell 2 in group 2 (since based on the current logic, it only splits based on test duration, and since each individual cell of each IPython notebook is treated as a single unit, they can get split into different groups). So, group 1 runs fine, but group 2 fails to run, not because there's something wrong with the notebook, but because it was split into 2 groups.

You can find all our notebooks [here](https://github.com/PMEAL/porespy/tree/dev/examples).

The reason I made this PR to your package, rather than forking is I assume `pytest-split` should should be test-agnostic, i.e. it should work on anything that's in principle testable using `pytest` itself. I understand if you want to keep the package minimal though, so no hard feelings 😄

**Comment by ma-sadeghi** [comment]
@jerry-git I just pushed a major refactor of the code I added, now down to ~10% of the entire code.

**Comment by jerry-git** [comment]
Thanks for the context and great use case for pytest-split 👍  

Alright, let's get the support done then.

File: src/pytest_split/plugin.py [hunk 4]
```diff
@@ -134,4 +134,23 @@ def _calculate_suite_start_and_end_idx(splits, group, items, stored_durations):
...
+        start_idx = _get_boundary_idx(item_node_ids, start_idx, "start")
+    if _is_ipy_notebook(item_node_ids[end_idx - 1]):  # since Python indexing is [a, b)
+        end_idx = _get_boundary_idx(item_node_ids, end_idx - 1, "end")
+
     return start_idx, end_idx
+
+
+def _is_ipy_notebook(node_id):
+    fpath = node_id.split("::")[0]
+    return True if fpath.endswith(".ipynb") else False
```
Comment (jerry-git) [inline src/pytest_split/plugin.py:148]: `return fpath.endswith(".ipynb")` should be enough

File: src/pytest_split/plugin.py [hunk 5]
```diff
@@ -134,4 +134,23 @@ def _calculate_suite_start_and_end_idx(splits, group, items, stored_durations):
...
+
+
+def _get_boundary_idx(item_node_ids, idx, mode):
```
Comment (jerry-git) [inline src/pytest_split/plugin.py:151]: As the mode is basically a flag (two possible values), I think boolean would be more typo-safe. E.g. `def _get_boundary_idx(item_node_ids, idx, *, start_mode)`

File: src/pytest_split/plugin.py [hunk 6]
```diff
@@ -134,4 +134,23 @@ def _calculate_suite_start_and_end_idx(splits, group, items, stored_durations):
     if not end_idx:
         end_idx = len(items)
 
+    # Modify start_idx and end_idx so they match start/end of an ipynb file
+    if _is_ipy_notebook(item_node_ids[start_idx]):
```
Comment (jerry-git) [inline src/pytest_split/plugin.py:138]: Could you please add some test case for this whole notebook thing?

Comment (jerry-git) [reply src/pytest_split/plugin.py:138]: Have a look at test_plugin.py. I think you should be able to create some simple test notebook with `testdir.makefile` and for the rest there should be good examples already in the existing tests.

Comment (ma-sadeghi) [reply src/pytest_split/plugin.py:145]: @jerry-git I added `test_ipynb.py` which is basically almost the same as `test_plugin.py`, but for notebooks. I tried merging it with `test_plugin.py` but it's my first time working with `pytest` fixtures, and I'm still trying to wrap my head around it! Also, for some reason, running pytest on `test_plugin.py` and `test_ipynb.py` individually works fine, but it fails when invoking it on both, i.e. via running simply `pytest`. Could you please take a look? Thanks.

**Comment by jerry-git** [comment]
Sorry it took a while to have a chance to look at this. It looks like there's something fishy going on with nbval, here's a minimal repro for the root cause of the issue:

```
def test_dummy1(example_suite):
    result = example_suite.inline_run("--nbval")
    assert len(_passed_test_names(result)) == 5

def test_dummy2(example_suite):
    result = example_suite.inline_run("--nbval")
    assert len(_passed_test_names(result)) == 5
```
The first one is successful, the second one fails.

So, basically, if I just run the exact same suite twice with only `--nbval` flag (i.e. no `pytest-split` related stuff happening), the second run fails. I'm not familiar with nbval so can't really say what's up without digging deeper. Have a look at how they are testing the nbval itself (https://github.com/computationalmodelling/nbval/tree/master/tests), maybe there's some explanation for this behaviour.

**Pull request closed without merging**